[et] commands for build and submitting Expo Go using EAS

wkozyra95 opened the pull request:
# Why

We want to use EAS Build & Submit to deliver Expo Go to the stores.

# How

- Remove existing workflows for building signed apps that were set up on a CI. I'm not sure if everything worked, plus we don't want to maintain both approaches.
- iOS Build & Submit:
  - download credentials from GCS bucket generated by Fastlane match
  - use openssl to convert credentials into p12 keystore
  - generate credentials.json
  - Run build and submit using EAS CLI
- Android Build&Submit
  - I don't have credentials yet so I used some random Keystore for testing. I'll download it from GCS bucket when it's ready
  - Create credentials.json.
  - Set track to internal, but it will be switched to production after everything is implemented and tested.
  - Run build and submit using EAS CLI. (Build failed for some unrelated reason, gha android builds are also failing)
- Enable remote version source. It can't be enabled per profile, so I set to 1 in `versioned-client`, `unversioned-client`, and `versioned-client-add-sdk` projects. For `release-client` I set  android version code to `189` (latest released version in playstore) and ios buildNumber to `1017524` (latest test flight version has `2.28.7.1017523`)
- Send notification to slack on successful build (before submit is started)

### Note on iOS versions

According to https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleversion?language=objc CFBundleVersion should be 3 numbers separated by a period, any additional values are ignored, so it looks like the values we have currently are not correct e.g. `2.28.7.1017523`.  As mentioned above I decided to change the build numbers to be a single integer and start with number that was on the last position in the latest test flight upload, from there new versions will be incremented

### TODO

- ~~Before merging this PR I'll set up all the missing credentials and verify everything e2e.~~
- In follow-up PRs, reimplement actions from Fastfile that are done during a release (disabling permissions, uploading crashlitics, maybe metadata update)
  - I don't plan to add version updates. Build number and version code will be updated automatically by EAS CLI, but user-facing version needs to be updated manually in xcode and android projects. If it's needed we can discuss adding some one place that will updated and sync rest of the project to that version

# Test Plan

Run new command on linux
TODO: test on macos

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

wkozyra95 committed:
release script

Kudo commented on tools/src/commands/EasDispatch.ts:
```diff
@@ -0,0 +1,176 @@
+    '-in',
+    certPEMPath,
+    '-password',
+    'pass:tmp-password',
```
nit: could we generate one-time random password rather than fixed one?

wkozyra95 replied:
sure, that is an easy change, although it's not making it any more secure. value stored in bucket is plain text public key, I only create keystore because there are some inconsistencies between PKCS keystore created with openssl and macOS keychain

Kudo commented on tools/src/commands/EasDispatch.ts:
```diff
@@ -0,0 +1,176 @@
+    'rsync',
+    '-r',
+    '-d',
+    'gs://expo-client-certificates',
```
nit: though it's not really critical. if we don't like exposing the bucket name, could we store the bucket name inside `secrets/keys.json` and let git-crypt help to do basic security for us?

wkozyra95 replied:
this is already public in fastlane/Matchfile

wkozyra95 commented on apps/eas-expo-go/eas.json:
```diff
@@ -91,6 +71,39 @@
+  },
+  "submit": {
+    "release-client": {
+      "ios": {
+        "ascApiKeyPath": "./credentials/asc_api_key.p8",
+        "ascApiKeyId": "GT3KPQKKBJ",
+        "ascApiKeyIssuerId": "69a6de96-160a-47e3-e053-5b8c7c11a4d1",
+        "ascAppId": "982107779"
+      },
+      "android": {
+        "serviceAccountKeyPath": "./credentials/api-8143610498327843040-246301-25839a6ccc53.json",
+        "track": "internal"
+      }
```
The other places you mentioned are relatively easy to change because they are in js, but here we would need to replace the file for a bit and restore it after. It's fine to do it if we need it, but my understanding is that all those ios values are harmless, and for Android I can change the path to sth without any ids.

Kudo replied:
thanks for following up on those comments. they are all about nit comments, it's not truly critical. please only address the comments only when those comments make sense to you. since the pr is about submitting expo-go, my idea is just that the less information we expose publicly, we increase more security.

all changes look good to me now. i'm approving this pr. if there are any other further changes, will take a look again.

tsapeta commented on apps/eas-expo-go/scripts/eas-build-pre-install.sh:
```diff
@@ -6,10 +6,20 @@ ROOT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"/../../.. && pwd )"
 if [ "$EAS_BUILD_PLATFORM" = "android" ]; then
   sudo apt-get -y update
-  sudo apt-get -y install ruby icu-devtools libicu66 libicu-dev maven
+  sudo apt-get -y install ruby icu-devtools libicu66 libicu-dev maven 
```
```suggestion
  sudo apt-get -y install ruby icu-devtools libicu66 libicu-dev maven
```

wkozyra95 committed:
send real slack notifications

brentvatne commented on apps/eas-expo-go/eas.json:
```diff
@@ -91,6 +71,40 @@
       }
+    },
+    "release-client": {
+      "extends": "versioned-client",
```
do we need to set `"autoIncrement": true` here? [given that the default is `false`](https://docs.expo.dev/build-reference/eas-json/#autoincrement)

wkozyra95 replied:
yes, I planned to add that after testing the full flow including submitting it to the store, there are also a few other change I mentioned in todos in a code
- add `--auto-submit`
- switch track to production for android

wkozyra95 committed:
finall changes after testing e2e

wkozyra95 committed:
revert version bump

wkozyra95 committed:
add some error handling to e.g. make it easier to debug gcloud issues

wkozyra95 committed:
do not add -legacy flag on macOS

wkozyra95 committed:
run bulid:version:sync after android build

wkozyra95 committed:
print info about EXPO_DEBUG only if EXPO_DEBUG is not set

wkozyra95 committed:
fix import

wkozyra95 commented:
- Added some error handling (realized that if someone would not have gcloud configured correctly it would be hard to debug). Everything is printed to terminal when EXPO_DEBUG=1 is set
- Running `eas build:version:sync` after android build (we don't need to run that after iOS because that version is not present in project sources)
- Tested on macOS and added fixes for openssl commands
